go/rep_subscription.go

```diff
@@ -0,0 +1,243 @@
+package pb
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+
+	libcryto "github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+//NewSignedSubscription creates a fully signed SignedSubscription with the given private key
+func NewSignedSubscription(topic string, version int64, r *Replication, key libcryto.PrivKey) (*SignedSubscription, error) {
+	ss := &SignedSubscription{}
+	ss.SetTopic(topic)
+	ss.SetVersion(version)
+	if err := ss.SetReplication(r); err != nil {
+		return nil, fmt.Errorf("error verifying replication: %s", err)
+	}
+	if err := ss.Sign(key); err != nil {
+		return nil, fmt.Errorf("error signing replication: %s", err)
+	}
+	return ss, nil
+}
+
+//GetTopic returns the topic of the SignedSubscription
+func (ss *SignedSubscription) GetTopic() string {
+	return ss.GetSubPart().GetTopic()
+}
+
+//SetTopic sets the topic of the SignedSubscription
+func (ss *SignedSubscription) SetTopic(topic string) {
+	ss.GetOrCreateSubPart().Topic = topic
+	ss.RemoveSignature()
+}
+
+//GetAuthorID returns the author of the Subscription
+func (s *Subscription) GetAuthorID() (peer.ID, error) {
+	return peer.IDFromBytes(s.AuthorIdBytes)
+}
+
+//GetAuthorID returns the author of the SignedSubscription, signature verification is not performed.
+func (ss *SignedSubscription) GetAuthorID() (peer.ID, error) {
+	return ss.GetSubPart().GetAuthorID()
+}
+
+//SetAuthorID sets the author of the Subscription
+func (s *Subscription) SetAuthorID(id peer.ID) error {
+	data, err := id.Marshal()
+	if err != nil {
+		return err
+	}
+	s.AuthorIdBytes = data
+	return nil
+}
+
+//SetAuthorID sets the author of the SignedSubscription. If success, any old signature is removed.
+func (ss *SignedSubscription) SetAuthorID(id peer.ID) error {
+	err := ss.GetOrCreateSubPart().SetAuthorID(id)
+	if err != nil {
+		return err
+	}
+	ss.RemoveSignature()
+	return nil
+}
+
+//GetVersion returns the version of the SignedSubscription, signature verification is not performed.
+func (ss *SignedSubscription) GetVersion() int64 {
+	return ss.GetUpdatePart().Version
+}
+
+//SetVersion sets the version of the SignedSubscription, and any old signature is removed.
+func (ss *SignedSubscription) SetVersion(v int64) {
+	ss.GetOrCreateUpdatePart().Version = v
+	ss.RemoveSignature()
+}
+
+//GetReplication return the replication file from this SubscriptionUpdate
+func (u *SubscriptionUpdate) GetReplication() (*Replication, error) {
+	if u == nil {
```

Comment:
Isn't this technically not needed? if `u` is nill, calling `u.GetReplication` would panic i think?

Reply:
Not with go typed nils. I added this during debuging to hunt down a nil pointer error.

You can see this style a lot in go protobuf generated code where a nil pointer returns default values. Good for certain coding styles, but can also hide a nil pointer error.

Reply:
Ah okay, makes sense